dkg/execution.go

```diff
@@ -0,0 +1,299 @@
+package dkg
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/drand/kyber"
+
+	"github.com/drand/drand/crypto"
+
+	"github.com/drand/drand/common"
+	"github.com/drand/drand/key"
+	"github.com/drand/drand/protobuf/drand"
+	"github.com/drand/drand/util"
+	"github.com/drand/kyber/share/dkg"
+	"github.com/drand/kyber/sign/schnorr"
+)
+
+func (d *DKGProcess) setupDKG(beaconID string) (*dkg.Config, error) {
+	current, err := d.store.GetCurrent(beaconID)
+	if err != nil {
+		return nil, err
+	}
+
+	lastCompleted, err := d.store.GetFinished(beaconID)
+	if err != nil {
+		return nil, err
+	}
+
+	keypair, err := d.beaconIdentifier.KeypairFor(beaconID)
+	if err != nil {
+		return nil, err
+	}
+	me, err := util.PublicKeyAsParticipant(keypair.Public)
+	if err != nil {
+		return nil, err
+	}
+
+	sortedParticipants := util.SortedByPublicKey(append(current.Remaining, current.Joining...))
+	var config *dkg.Config
+	if lastCompleted == nil {
+		config, err = d.initialDKGConfig(current, keypair, sortedParticipants)
+	} else {
+		config, err = d.reshareDKGConfig(current, lastCompleted, keypair, sortedParticipants)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// create the network over which to send all the DKG packets
+	board, err := newEchoBroadcast(
+		d.internalClient,
+		d.log,
+		common.GetAppVersion(),
+		beaconID,
+		me.Address,
+		sortedParticipants,
+		keypair.Scheme(),
+		config,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// we need some state on the DKG process in order to process any incoming gossip messages from the DKG
+	// if other nodes try to send us DKG messages before this is set we're in trouble
+	d.lock.Lock()
+	d.Executions[beaconID] = board
+	d.lock.Unlock()
+
+	return config, nil
+}
+
+// this is done rarely and is a shared object: no good reason not to use a clone (and it makes the race checker happy :))
+//
+//nolint:gocritic
+func (d *DKGProcess) executeAndFinishDKG(beaconID string, config dkg.Config) error {
+	current, err := d.store.GetCurrent(beaconID)
+	if err != nil {
+		return err
+	}
+
+	lastCompleted, err := d.store.GetFinished(beaconID)
+	if err != nil {
+		return err
+	}
+
+	executeAndStoreDKG := func() error {
+		output, err := d.startDKGExecution(beaconID, current, &config)
+		if err != nil {
+			return err
+		}
+
+		finalState, err := current.Complete(output.FinalGroup, output.KeyShare)
+		if err != nil {
+			return err
+		}
+
+		d.completedDKGs <- SharingOutput{
+			BeaconID: beaconID,
+			Old:      lastCompleted,
+			New:      *finalState,
+		}
+
+		err = d.store.SaveFinished(beaconID, finalState)
```

Comment:
Better to save locally first before sending ot channel in case channel is full or not listened on or stg ?